### PR TITLE
roachtest: default Run{,E} to a per-invocation file

### PR DESCRIPTION
**Pull request by tbg**
A regular complaint about roachtest is that it is difficult to reason about
the flow of the test given only the log files. This is in part because the
main log often contains lenghty workload outputs that create large files
with lots of non-timestamped sections where output is often even hard to
assign to a specific process. Questions that often come up are "did this
process already terminate at this point?", "Did it ever print output?",
"What process does this log line belong to", "What is this test even doing?
All I'm seeing are pages and pages of some bogus workload output",among
many others.

This commit changes the default behavior to creating a unique file per
invocation (sorting in timestamp order) and streaming the output there and
only there. Where streaming to stdout/stderr is desired, `RunL` can be used
instead.

Certainly a lot more can be done here, but this could already help quite a
bit in keeping the main log manageable in size while also allowing us to
better diagnose test failures.

Release note: None

### CODE REVIEW

File: pkg/cmd/roachtest/cluster.go [hunk 1]
```diff
@@ -1806,7 +1806,25 @@ func (c *cluster) Install(
...
+	cmdString := strings.Join(args, " ")
+	logFile := fmt.Sprintf(`run_%s_%s_%s`,
+		timeutil.Now().Format(`15:04:05.000`),
+		c.makeNodes(node),
+		regexp.MustCompile(`[^a-z0-9_]+`).ReplaceAllString(cmdString, "_"),
+	)
+	// NB: we set no prefix because it's only going to a file anyway.
+	l, err := c.l.ChildLogger(logFile, quietStderr, quietStdout)
+	if err != nil {
+		return c.RunL(ctx, c.l, node, args...)
```
Comment (tbg) [inline pkg/cmd/roachtest/cluster.go]: I should probably just return the error here. Is it ever expected that ChildLogger fails?

**Review by andreimatei — COMMENTED** [review]
For the 2nd commit, I think it's great to cut down the number of logger instantiations, but I like passing the  loggers explicitly to various methods....

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43129#-:-LvvCIRR8ftkOfRpwU4m:b-16l8yi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @tbg)

___
*[pkg/cmd/roachtest/cluster.go, line 1807 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43129#-LvvBImj43TdSdmdM0FT:-LvvBImk59GDP3oMasvH:b-roz6wz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2a1451c06665452c8f1334ff2151f94a655f6af/pkg/cmd/roachtest/cluster.go#L1807)):*
> ```Go
> }
> 
> // RunE runs a command on the specified node, returning an error.
> ```

please add a comment about where the log goes
___
*[pkg/cmd/roachtest/cluster.go, line 1811 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43129#-LvvBajLEAPgvl9duAfo:-LvvBajLEAPgvl9duAfp:bl33amy) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2a1451c06665452c8f1334ff2151f94a655f6af/pkg/cmd/roachtest/cluster.go#L1811)):*
> ```Go
> 	dest := fmt.Sprintf(`run_%s_%s_%s`,
> 		timeutil.Now().Format(`15:04:05.000`),
> 		c.makeNodes(node),
> ```

what's the point of this `makeNodes()` call?
___
*[pkg/cmd/roachtest/cluster.go, line 1818 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43129#-LvuNlh4QSl5gVOktuk3-r2-1818:-LvvBrzTAg-_kCFcvATO:b-ejt591) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2d43236c1f53443e22b9bc122c5535d1c3df85c/pkg/cmd/roachtest/cluster.go#L1818)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I should probably just return the error here. Is it ever expected that ChildLogger fails?
</blockquote></details>

return err indeed

**Review by andreimatei — COMMENTED** [review]
In fact, I hope to get rid of `c.l`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43129#-:-LvvCVKr1znz8AbJ5hUt:bh7gcnh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @tbg)

**Review by tbg — COMMENTED** [review]
TFTR

For the record, Andrei and I chatted offline and I got an LGTM. We are interested in cleaning up the interplay between `c.l`, `c.test`, `test.l` to make sure we get rid of all of this confusion, but this is out of scope here.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43129#-:-Lyctaae0Xhql4vjUwAr:bs4z06r)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/cmd/roachtest/cluster.go, line 1807 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43129#-LvvBImj43TdSdmdM0FT:-Lyct_pKDmKdS1pzIlOf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2a1451c06665452c8f1334ff2151f94a655f6af/pkg/cmd/roachtest/cluster.go#L1807)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please add a comment about where the log goes
</blockquote></details>

Done.
___
*[pkg/cmd/roachtest/cluster.go, line 1811 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43129#-LvvBajLEAPgvl9duAfo:-Lycsnrc6IxeQVFz6gD8:b-en831e) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2a1451c06665452c8f1334ff2151f94a655f6af/pkg/cmd/roachtest/cluster.go#L1811)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

what's the point of this `makeNodes()` call?
</blockquote></details>

It makes the filename describe the full command. Without this, the filename wouldn't reflect where the command was run. For example

`c.Run(ctx, c.Range(1,3), "echo", "foo")` would create some file `15_04_05.000_echo_foo`, but the actual code will make it `15_04_05.000_1-3_echo_foo`. I think this will be helpful sometimes (and it doesn't hurt if it's not).

**Comment by tbg** [comment]
bors r=andreimatei

Getting very interested in merging this again while looking at all of the tests in https://github.com/cockroachdb/cockroach/issues/43754

**Comment by tbg** [comment]
CI failed because the created filenames could get too long. I made them more succinct (and generally brushed this PR up).

**Comment by tbg** [comment]
bors r=andreimatei

**Pull request merged**